### ISSUE TITLE
GHC #4092 is closed

### ISSUE THREAD

**Issue by int-index** [well-typed/cborg#173]
The code has a reference to Trac ticket 4092, which is closed.

https://github.com/well-typed/cborg/blob/99ec2b038538b279625c10640e3d5e8d80cd7a1d/cborg/src/Codec/CBOR/Magic.hs#L300

This probably means that the relevant piece of code can take advantage of the added feature.

**Comment by thoughtpolice** [comment]
Here's a patch for this. The code actually isn't improved, the difference between this patch being applied or not on 8.4 seems to be noise; I assume it might be because the stupid benchmark I wrote for floats is just bad. I'll try to make a better one if I can...

```diff
diff --git a/cborg/src/Codec/CBOR/Magic.hs b/cborg/src/Codec/CBOR/Magic.hs
index c2900db..ff1adc5 100644
--- a/cborg/src/Codec/CBOR/Magic.hs
+++ b/cborg/src/Codec/CBOR/Magic.hs
@@ -115,6 +115,10 @@ import           GHC.IntWord64 (wordToWord64#, word64ToWord#,

 #endif

+#if defined(HAVE_OPTIMIZED_FLOAT_CONVERSIONS)
+import           GHC.Float     (castWord32ToFloat, castWord64ToDouble)
+#endif
+
 --------------------------------------------------------------------------------

 -- | Grab a 8-bit @'Word'@ given a @'Ptr'@ to some address.
@@ -307,14 +311,25 @@ floatToWord16 = \x -> cast (Half.getHalf (Half.toHalf x))

 -- | Cast a @'Word32'@ to a @'Float'@.
 wordToFloat32 :: Word32 -> Float
-wordToFloat32 (W32# w#) = F# (wordToFloat32# w#)
 {-# INLINE wordToFloat32 #-}

 -- | Cast a @'Word64'@ to a @'Float'@.
 wordToFloat64 :: Word64 -> Double
-wordToFloat64 (W64# w#) = D# (wordToFloat64# w#)
 {-# INLINE wordToFloat64 #-}

+
+#if defined(HAVE_OPTIMIZED_FLOAT_CONVERSIONS)
+wordToFloat32 w = castWord32ToFloat w
+wordToFloat64 w = castWord64ToDouble w
+#else
+wordToFloat32 (W32# w#) = F# (wordToFloat32# w#)
+wordToFloat64 (W64# w#) = D# (wordToFloat64# w#)
+#endif
+
+--
+-- Primitives, used on GHC 8.2 and below
+--
+
 -- | Cast an unboxed word to an unboxed float.
 wordToFloat32# :: Word# -> Float#
 wordToFloat32# w# =
diff --git a/cborg/src/cbits/cbor.h b/cborg/src/cbits/cbor.h
index f2cd151..384f12d 100644
--- a/cborg/src/cbits/cbor.h
+++ b/cborg/src/cbits/cbor.h
@@ -9,6 +9,14 @@
 #define HAVE_BYTESWAP_PRIMOPS
 #endif

+/*
+** GHC 8.4 and above include better primitive operations for casting betweeen
+** word/float values, see https://ghc.haskell.org/trac/ghc/ticket/4092
+*/
+#if __GLASGOW_HASKELL__ >= 804
+#define HAVE_OPTIMIZED_FLOAT_CONVERSIONS
+#endif
+
 /*
 ** On Intel 32/64 bit machines, memory access to unaligned addresses
 ** is permitted (and generally efficient, too). With this in mind,
```

**Comment by bgamari** [comment]
I believe this was resolved in #184.
